tests/queries/0_stateless/02456_test_attach_partition_different_partition_exp.sql

```diff
@@ -0,0 +1,154 @@
+-- { echoOn }
+-- Should be allowed since destination partition expr is monotonically increasing and compatible
+DROP TABLE IF EXISTS source;
+DROP TABLE IF EXISTS destination;
+
+CREATE TABLE source (timestamp DateTime) engine=MergeTree ORDER BY tuple() PARTITION BY toYYYYMMDD(timestamp);
+CREATE TABLE destination (timestamp DateTime) engine=MergeTree ORDER BY tuple() PARTITION BY toYYYYMM(timestamp);
```

Comment:
And a test with `partition by (col1, col2)`

Reply:
from `tuple()` to `toYYYYMM(timestamp)`? I guess it makes sense, but currently does not work. I think I'll have to work on that

Reply:
for `partition by (col1, col2)` it will not work because it'll consider it not be monotonic (2 columns, no constant).

Reply:
> from tuple() to toYYYYMM(timestamp)?

I'll see if I can get it working.

> partition by (col1, col2) to partition by col1

The initial requirements I came up with was that the columns should match, but maybe I can relax that to `source_partition_expression_columns.contains(destination_partition_expression_columns) == true`. 

> partition by (col1 / 2, col2 / 2) to partition by (col1, col2)

Current implementation of monotonicity check works for functions of a single argument or two arguments as long as one of them is a constant. It's already implemented like this in `KeyCondition` and even more strict in `MonotonicityCheckVisitor`. Maybe it can be implemented in the future?

Reply:
It's not a function of two arguments, it's two independent functions of one argument. At first glance, we need to check that both of them are monotonic, but I'm not sure.

Reply:
> At first glance, we need to check that both of them are monotonic, but I'm not sure.

Btw, that's similar to how `MonotonicityCheckVisitor` is used for `ORDER BY` expression optimization: [link](https://github.com/ClickHouse/ClickHouse/blob/6c9a1b14ef9cc9c1fec3aec1c8ace34c889942ec/src/Interpreters/TreeOptimizer.cpp#L431-L451)

Reply:
@tavplubix @alexey-milovidov 

Implemented attach from `partition by toYYYYMM(timestamp)` to `partition by tuple()` (i.e, from partitioned to unpartitioned). I had to add a workaround for `partition by tuple()` monotonicity check. Please let me know if there is a better way to do that.

The other way around, attach from unpartitioned to partitioned is a bit tricky. My implementation relies on min max indexes to make it a fast operation. If the source partition is unpartitioned, it might not contain the minmax indexes pre-calculated. IMO, we should skip this use case for now.

What do you think?

Reply:
> The other way around, attach from unpartitioned to partitioned is a bit tricky.

Not supporting this is okay for now

Reply:
@tavplubix I believe the logic will need some short-circuits and special checks to address a few corner cases.

Examples:

1. `partition by (a, b, c)` to `partition (a, b)` - we don't need to check for monotonicity, destination expression is a "direct" subset of source expression.
2.  `partition by toYYYYMM(timestamp)` to `partition by tuple()` or without `partition by` clause - we don't need to check for monotonicity, there is only one partition in the destination table.
3. Other cases will have to match the following requirements:
    1. Columns required by destination partition expression must be a subset of source partition expression.
    2. Destination partition expression must contain at most 2 arguments. In case the number of arguments is 2, one of them must be a constant.
    3. Destination partition expression is monotonically increasing on the source partition min max range.
    4. All data parts would fall into the same partition.

Being that said, the cases that are **NOT** supported:

1. `partition by (col1 / 2, col2 / 2)` to `partition by (col1, col2)` - It is not a direct subset and function contains two variables. I know in practice this would be compatible, but it is hard to check.
2. unpartitioned to partitioned as discussed above.

Reply:
@tavplubix 

Just finished implementing attach to partition expression with multiple expressions that are functions of at most 2 arguments, one of them being a constant. Example: `PARTITION BY tuple(A, B)` to `PARTITION BY tuple(intDiv(A, 2), intDiv(B, 2))`

Reply:
> Please add at least one test with `ReplicatedMergeTree` as well

It does not work as expected, it looks like I need to change some stuff in https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/StorageReplicatedMergeTree.cpp#L7762.

Thanks for bringing this up. Do you know any other similar case that should be tested?

Reply:
Added a draft implementation for `ReplicatedMergeTree`, will dive deeper into it tomorrow

Reply:
Added integration tests for `ReplicatedMergeTree`